SteamKit2/SteamKit2/Steam/CDN/Client.cs

```diff
@@ -227,10 +232,31 @@ public async Task<int> DownloadDepotChunkAsync( uint depotId, DepotManifest.Chun
                 }
             }
 
+            await _lancacheDetectionLock.WaitAsync();
+            try
+            {
+                if ( !CheckedForLancacheServer )
+                {
+                    LancacheDetected = await LancacheDetector.DetectLancacheServerAsync(httpClient);
+                }
+            }
+            finally
+            {
+                CheckedForLancacheServer = true;
+                _lancacheDetectionLock.Release();
+            }
+
             var chunkID = Utils.EncodeHexString( chunk.ChunkID );
             var url = $"depot/{depotId}/chunk/{chunkID}";
 
-            using var request = new HttpRequestMessage( HttpMethod.Get, BuildCommand( server, url, cdnAuthToken, proxyServer ) );
+            var builtUrl = BuildCommand( server, url, cdnAuthToken, proxyServer );
+            using var request = new HttpRequestMessage( HttpMethod.Get, builtUrl );
+            if ( LancacheDetected )
+            {
+                request.Headers.Host = server.Host;
```

Comment:
This looks like a bit of a dodgy approach to me.

I think it would be better to override the [ConnectCallback](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-8.0) so that the HTTP request is exactly the same as it would be to the CDN, we just connect to the lancache server instead of the CDN.

That would also help minimize the amount of lancache-specific branching logic here.

Reply:
I'm not very familiar with ConnectCallback so I'm not exactly sure how to implement this change using it.  Maybe I can provide some more details on why I implemented it this way and you can direct me where to start. 

The lancache instance is ultimately just an nginx server with some game cdn specific config. For pretty much every other cdn we just poison the appropriate domain names, and nginx will automatically forward upstream requests to the correct server.  In this case the client applications have no idea that anything is running through the nginx cache. 
Steam however is different in that they provided us a special domain `lancache.steamcontent.com` which doesn't **technically** point to one of their cdns, it's expected to always resolve to a local ip if there is a cache present.  Because of there being no actual upstream at `lancache.steamcontent.com` Nginx will need to know which upstream server to use instead, which is specified in the required Host header.  The other piece of the puzzle is that the user agent must absolutely match the one that Steam uses, otherwise requests won't complete successfully. 

As long as we can make sure those two headers are set correctly then we should be good to go with the implementation you're suggesting

Reply:
ConnectCallback should allow us to replace the IP it connects to (the local one in this case), without manually changing the host header.

Reply:
If we implement pointing to the lancache ip in ConnectCallback we will still need to pass the upstream cdn in the host header, otherwise the cache will just fail the request. Since we no longer have the original domain in the request (just the ip) nginx will need to be explicitly told what the upstream is.   I can show some example requests that demonstrate this behavior if you'd like.